### ISSUE TITLE
CRAN/BIOC older versions (re-add)

### ISSUE THREAD

**Issue by gaborcsardi** [r-lib/pkgdepends#110]

**Comment by gaborcsardi** [comment]
Yes, this would be great, but not really feasible currently in my opinion.

Collecting the metadata is one issue. It is actually available at crandb.r-pkg.org, e.g. https://crandb.r-pkg.org/cli/all are all cli versions, but this is per package, and downloading updates efficiently is tricky. It is also not completely in sync with the various CRAN mirrors, which is another (smaller) issue.

A more serious issue is that you cannot simply put all these package versions into the dependency solver, because you'll get a combinatorial explosion. We would need heavy heuristics to eliminate package versions from the solution space. E.g. if package X 1.0 and the its next 1.1 version have the same exact dependencies, then we can eliminate 1.0, there is not reason to install that.

I hope that we would have this one day, but it won't be soon.

This issue is about a much more modest feature, being able to request one specific version of a CRAN package, basically pinning a version, without constraining the rest of the packages. Admittedly, this often fails, e.g. if the requested old version depends on a package that is not on CRAN any more. With this you could have at least a relatively straightforward solution to the problem you describe above.

**Comment by dgkf** [comment]
I see - thanks for the clarification. 

I didn't know `crandb` exposed this info - I'll do some exploration to see how this could be used. 

This solution would only apply to `CRAN`, correct? A `crandb` analog would be needed for each endpoint in `repos`?

Assuming this would blow up the problem space on an ILP problem, are there other solver strategies that you are considering that would address this?

**Comment by gaborcsardi** [comment]
Right, every repo would need to have this metadata in some form. Not necessarily like crandb, a big PACKAGES file with all versions might be better.

Re dependency solvers, you can formalise it as a SAT problem for example, and some people found that SAT solvers scale better on these kinds of problems. But changing the solver is quite some work, and at the end of the day it is an NP-complete problem, so nothing will scale well enough to be able to include 5-10+ versions for every package.

**Comment by gaborcsardi** [comment]
Btw. another thing to consider is that because of the rolling nature of CRAN only certain combinations of package versions are tested together. So just because package A did not require a certain version of its package B dependency, that does not mean that they work well together. Many packages don't declare their dependencies, and packages don't use semantic versioning, so the usefulness of a solver that considers all versions is questionable.

Maybe we could have a compromise and consider the last two versions of every package. That would solve many issues in practice and it is easier to implement.

**Comment by dgkf** [comment]
> because of the rolling nature of CRAN only certain combinations of package versions are tested together. So just because package A did not require a certain version of its package B dependency, that does not mean that they work well together

Yes - I totally agree and understand that the utility starts to fade. 

That's sort of the reason why this problem piqued my interest. My theory is that being more conscious of dependency constraints is under-incentivized because the only cohorts that offer any guarantees are snapshots of CRAN. If R had more comprehensive tools for resolving dependencies (perhaps via `pkgdepends`) beyond the current CRAN snapshot, I'd hope that it becomes more useful to be prescriptive about valid dependency versions.

Perhaps that's too idealist, though - and I agree that given the way the ecosystem works today, it wouldn't add a lot of value. Just considering the last two versions might be enough for most situations.
